pylib_3.9.7/df/ProteinRASA.py

```diff
@@ -0,0 +1,148 @@
+from df.data_transfer import ColumnData, TableData, DataFunctionRequest, DataFunctionResponse, DataFunction, DataType, \
+                             string_input_field, boolean_input_field, integer_input_field,  input_field_to_column, \
+                             Notification, NotificationLevel
+
+from ruse.bio.antibody import align_antibody_sequences, NumberingScheme, CDRDefinitionScheme, ANTIBODY_NUMBERING_COLUMN_PROPERTY
+from ruse.bio.bio_data_table_helper import sequence_to_genbank_base64_str
+
+class ProteinRASA(DataFunction):
+    """
+    Computes relative solvent accessible surface area of a given protein
+    """
+
+    def execute(self, request: DataFunctionRequest) -> DataFunctionResponse:
+
+        # get settings from UI
+        structure_column = input_field_to_column(request, 'uiStructureColumn')
+        structures, notifications = column_to_structures(structure_column)
+
+        sequence_column = input_field_to_column(request, 'uiSequenceColumn')
+        sequences = column_to_sequences(sequence_column)
+
+        rasa_cutoff = integer_input_field(request, 'uiRASACutoff')
+        exposedColor = string_input_field(request, 'uiExposedColor')
+
+        labelAll = boolean_input_field(request, 'uiLabelAll')
+        startColor = string_input_field(request, 'uiStartColor')
+        endColor = string_input_field(request, 'uiEndColor')
+
+        if labelAll:
+            RASA_color_gradient = generate_color_gradient(startColor, endColor, 21)
+
+        # setup BioPython tools
+        sr = ShrakeRupley()
+        backbone_atoms = ['N', 'H', 'CA', 'HA', 'C', 'O', 'H2', 'H3']
+
+        # output storage
+        output_sequences = []
+
+        # process each structure
+        for index, (structure, sequence) in enumerate(zip(structures, sequences)):
+            if not (structure and sequence):
+                notifications.append(Notification(level = NotificationLevel.ERROR,
+                                                  title = 'Relative Accessible Surface Area',
+                                                  summary = f'Error for structure or sequence in Row {index}/nNull values found.',
+                                                  details = f'Either the structure or sequence were null.'))
+                continue
+
+            # create a mapping between the structure residues and potentially gapped sequence residues
+            # would this be a useful biotools function rather than buried here?
+            sequence_number_mapping = {}
+            structure_residue_number = 0
+            structure_residues = [amino3to1(res.get_resname()) for res in structure.get_residues()]
+            for sequence_residue_number, sequence_residue in enumerate(sequence.seq):
+                if sequence_residue != '-':
+                    if sequence_residue != structure_residues[structure_residue_number]:
+                        # send a notification and then fail
```

Comment:
If you raise an error here that will fail the data function and create a Spotfire notification